src/OpenTK/Platform/Linux/LinuxJoystick.cs

```diff
@@ -71,29 +74,144 @@ internal sealed class LinuxJoystick : IJoystickDriver2
-                watcher.Path = path;
+                OpenJoysticks(EvdevPath);
+            }
+        }
+
+        private MD5 Hasher = MD5.Create();
 
-                watcher.Created += JoystickAdded;
-                watcher.Deleted += JoystickRemoved;
-                watcher.EnableRaisingEvents = true;
+        private string procHash;
 
-                OpenJoysticks(path);
+        private void ProcessEvents()
+        {
+            if (!File.Exists(InputDevicePath))
+            {
+                //BSD may not have the Linux ProcFS, no use running a pointless loop
+                return;
+            }
+
+            while (!disposed)
+            {
+                /*
+                 * Rather hacky filesystemwatcher equivilant, as it doesn't work on ProcFS
+                 * As the thing is psuedo-files, we can't use the last modified date either
+                 * We have to load the thing into a stream and hash. At least it's only a few bytes.....
+                 *
+                 * Note: We could possibly override the Mono filesystem watcher backend, but this is a global variable only
+                 * and so may break user code
+                 */
+                List<int> connectedDevices = new List<int>();
+                using (StreamReader reader = new StreamReader(new FileStream(InputDevicePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite)))
+                {
+                    string newHash = BitConverter.ToString(Hasher.ComputeHash(reader.BaseStream));
+                    if (newHash == procHash)
+                    {
+                        //Marginal goto, but might as well only open the stream once
+                        goto EndCondition;
+                    }
+
+                    reader.BaseStream.Seek( 0, 0);
+                    procHash = newHash;
+                    OpenJoysticks(EvdevPath);
+                    while (!reader.EndOfStream)
+                    {
+                        string currentLine = reader.ReadLine();
+                        if (string.IsNullOrEmpty(currentLine))
+                        {
+                            continue;
+                        }
+
+                        switch (char.ToLower(currentLine[0]))
+                        {
+                            case 'h':
+                                string[] handlers = currentLine.Substring(currentLine.IndexOf('=') + 1).Trim().Split(' ');
+
+                                foreach (string handler in handlers)
+                                {
+                                    if (handler.StartsWith("event", StringComparison.InvariantCultureIgnoreCase))
+                                    {
+                                        //As device is in /proc/bus/input/devices it is connected, so pull out the evdev ID
+                                        int evdevID = int.Parse(handler.Substring(5));
+                                        connectedDevices.Add(evdevID);
```

Comment:
Is this collection exposed anywhere? This might trigger an exception if we add to it during iteration.

Reply:
Not exposed, purely an internal for use when updating.
Could make it global and clear instead of creating a new one but IIRC this does much the same in the backing code.